encfs/compatwin.cpp

```diff
@@ -665,31 +687,33 @@ unix::stat(const char *path, struct stat_st *buffer)
 
 
   unsigned mode;
-  if (wfd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+  if ((hfi.dwFileAttributes + wfd.dwFileAttributes) & FILE_ATTRIBUTE_DIRECTORY)
     mode = _S_IFDIR | 0777;
   else
     mode = _S_IFREG | 0666;
   // Set attributes of file/directory
-  if (wfd.dwFileAttributes & FILE_ATTRIBUTE_READONLY)
+  if ((hfi.dwFileAttributes + wfd.dwFileAttributes) & FILE_ATTRIBUTE_READONLY)
     mode &= ~0222;
-
+  // The following solution is not complete, Cygwin does not correctly detect such items as links...
+  // if ((hfi.dwFileAttributes + wfd.dwFileAttributes) & FILE_ATTRIBUTE_REPARSE_POINT)
+  //   mode |= S_IFLNK;
 
   buffer->st_dev = buffer->st_rdev = drive;
-  buffer->st_ino = wfd.nFileIndexHigh * (((uint64_t)1) << 32) + wfd.nFileIndexLow;
+  buffer->st_ino = (hfi.nFileIndexHigh + 0) * (((uint64_t)1) << 32) + (hfi.nFileIndexLow + 0);
```

Comment:
I'm curious what the impacts will be since FindFirstFileW does not return the file ID ... I believe this is important for readdir.

Reply:
In this case `st_ino` will be set to 0.
Strangely enough, `ls -li` with Cygwin does show an inode number for the files stated by `FindFirstFileW`. Not sure where is exactly comes from.

In EncFS, inode number is used in reverse mode to compute the IV (when IV is enabled).
This would then lead to duplicate IV for the considered files.
However, these files were stated by `FindFirstFileW` because produced a permission denied with `GetFileInformationByHandle`.
So EncFS will not be able to read them... So I think duplicate IV should not be an issue 👍

Reply:
Would it be enough to return an errno of ENOENT or EACCES if GetFileInformationByHandle fails, since encfs won't be able to use the file anyways?  Or is this not enough to fix the issue in #98?  Of course in this case any files that cause GetFileInformationByHandle to fail due to access restrictions would not show up at all in reverse mode ...

Reply:
Problem is that if `stat` function returns an error for one of a directory elements, directory listing stops immediately. This is actually the root cause of #98. You can easily reproduce this.
This is why `FindFirstFileW` comes to the rescue, as it manages to stat problematic files, thus making directory listing to complete.
I also think that having missing items in a directory listing could be a problem. Think for example about a backup program which would not see such files, and then would not return any issue to the backup admin who would perhaps have expected them to be backed-up...

Reply:
Yeah it looks like when stat returns an error it causes [nextPlaintextName](https://github.com/jetwhiz/encfs4win/blob/master/encfs/DirNode.cpp#L100) to return an empty string, causing [encfs_readdir](https://github.com/jetwhiz/encfs4win/blob/master/encfs/encfs.cpp#L255) to terminate prematurely (thinking there are no files left).  

I'm guessing this isn't an issue upstream since Linux doesn't run into these errors like with Windows when `stat`ing, since Windows needs to actually open a filehandle to get all of its attributes?  `FindFirstFileW` does not need a filehandle, but cannot give us the file ID unfortunately. 

I'm wondering if there are any potential race conditions though if files are inaccessible only while performing the readdir (if a user has the files opened, for instance), but then are accessible when encfs later tries to do something with the file.  I'm guessing these files would have `stat` re-performed before encfs would do anything, though (giving them the correct file ID).

Reply:
After further investigation, I found that the culprit is this the [`fuse_fill_dir_t filler()` call](https://github.com/jetwhiz/encfs4win/blob/v1.10.1-RC13/encfs/encfs.cpp#L264) in `encfs_readdir()`. Internally it calls `stat` on each entry to get further info, and returns false when it fails, thus making the loop to break.

I'm thinking about pushing a patch to upcoming EncFS 1.9.5 to avoid breaking the loop, but printing error messages instead. We would still need `FindFirstFileW` though, so not sure it's really needed. Perhaps to avoid cases where `FindFirstFileW` would not be able to stat ?
Any thoughts ?

Regarding races and the inode number computed during readdir, see my last commit : https://github.com/jetwhiz/encfs4win/pull/103/commits/1b1f962b78c9ae2f92756cd0531cdfecf8a566b8
When inode number will be needed, it will be retrieved using `stat` function :
https://github.com/jetwhiz/encfs4win/blob/v1.10.1-RC13/encfs/CipherFileIO.cpp#L274